### PR TITLE
Fix ethereum subscriptions

### PR DESCRIPTION
**Pull request by KillariDev**
Fix https://github.com/DarkFlorist/TheInterceptor/issues/302

we also reorg the simulated block on top when new block arrives

### CODE REVIEW

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 1]
```diff
@@ -359,8 +359,16 @@ const getBaseFeePerGasForNewBlock = (parent_gas_used: bigint, parent_gas_limit:
...
-	if (simulationState == undefined || await canQueryNodeDirectly(ethereumClientService, simulationState, blockTag)) return await ethereumClientService.getBlock(blockTag, fullObjects)
+export async function getSimulatedBlock(ethereumClientService: EthereumClientService, simulationState: SimulationState, blockTag: EthereumBlockTag, fullObjects: true): Promise<EthereumBlockHeader>
+export async function getSimulatedBlock(ethereumClientService: EthereumClientService, simulationState: SimulationState, blockTag: EthereumBlockTag, fullObjects: false): Promise<EthereumBlockHeaderWithTransactionHashes>
+export async function getSimulatedBlock(ethereumClientService: EthereumClientService, simulationState: SimulationState, blockTag: EthereumBlockTag = 'latest', fullObjects: boolean = true): Promise<EthereumBlockHeader | EthereumBlockHeaderWithTransactionHashes>  {
+	if (simulationState == undefined || await canQueryNodeDirectly(ethereumClientService, simulationState, blockTag)) {
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:365]: Never 2, always 3.
```suggestion
	if (simulationState === undefined || await canQueryNodeDirectly(ethereumClientService, simulationState, blockTag)) {
```

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 2]
```diff
@@ -359,8 +359,16 @@ const getBaseFeePerGasForNewBlock = (parent_gas_used: bigint, parent_gas_limit:
...
+		if (fullObjects) {
+			return await ethereumClientService.getBlock(blockTag, true)
+		} else {
+			return await ethereumClientService.getBlock(blockTag, false)
+		}
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:370]:
```suggestion
		return await ethereumClientService.getBlock(blockTag, !!fullObjects)
```

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 3]
```diff
@@ -385,22 +393,14 @@ export const getSimulatedBlock = async (ethereumClientService: EthereumClientSer
 		totalDifficulty: parentBlock.totalDifficulty + parentBlock.difficulty, // The difficulty increases about the same amount as previously
 		uncles: [],
 		baseFeePerGas: getBaseFeePerGasForNewBlock(parentBlock.gasUsed, parentBlock.gasLimit, parentBlock.baseFeePerGas),
-		transactionsRoot: parentBlock.transactionsRoot // TODO: this is wrong
+		transactionsRoot: parentBlock.transactionsRoot, // TODO: this is wrong
+		transactions: simulationState.simulatedTransactions.map((simulatedTransaction) => {
+			return simulatedTransaction.signedTransaction
+		})
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:399]: Style preference:
```suggestion
		transactions: simulationState.simulatedTransactions.map((simulatedTransaction) => simulatedTransaction.signedTransaction)
```

**Review by MicahZoltu — APPROVED** [review]

File: extension/app/ts/simulation/services/EthereumSubscriptionService.ts [hunk 4]
```diff
@@ -1,92 +1,75 @@
...
+	console.log('createsub tabid', subscriptionCreatorSocket.tabId)
+	switch(params.params[0]) {
+		case 'newHeads': {
+			const subscriptionId = generateId(40)
+			await updateEthereumSubscriptions((subscriptions: EthereumSubscriptions) => {
+				return subscriptions.concat({ type: 'newHeads', subscriptionId, params, subscriptionCreatorSocket })
+			})
+			return subscriptionId
+		}
+		case 'logs': throw `Dapp requested for 'logs' subscription but it's not implemented` //TODO: implement
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/simulation/services/EthereumSubscriptionService.ts:71]: hmm, we need to remove re-orged logs?

[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/simulation/services/EthereumSubscriptionService.ts:71]: I wonder if there's a dapp that would use this feature? I haven't seen one and thats why they are not implemented

File: extension/app/ts/background/simulationModeHanders.ts [hunk 5]
```diff
@@ -17,8 +17,10 @@ import { openPersonalSignDialog } from './windows/personalSign.js'
 const defaultCallAddress = 0x1n
 
 export async function getBlockByNumber(ethereumClientService: EthereumClientService, simulationState: SimulationState, request: EthBlockByNumberParams) {
-	const block = await getSimulatedBlock(ethereumClientService, simulationState, request.params[0], request.params[1])
-	return { result: GetBlockReturn.serialize(block) }
+	if (request.params[1] === true) {
+		return { result: GetBlockReturn.serialize(await getSimulatedBlock(ethereumClientService, simulationState, request.params[0], true)) }
+	}
+	return { result: GetBlockReturn.serialize(await getSimulatedBlock(ethereumClientService, simulationState, request.params[0], false)) }
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/background/simulationModeHanders.ts:23]: !! feels quite ugly and hackish, I feel the above is more clear?

**Pull request merged**